test/unittest.html

```diff
@@ -10,6 +10,22 @@
         <div id="messages"></div>
         <div id="fixtures"></div>
 
+        <script src="../node_modules/sinon/pkg/sinon.js"></script>
+        <script src="../node_modules/fetch-mock/es5/client-bundle.js"></script>
+        <script>
+            /* 
+             * In order to mock requests they must be setup here before the tracer
+             * is loaded. This is because the request plugins grab fetch/xhr immediately 
+             * so the mocks must be initialized before that happens. xhr requests are
+             * mocked by sinion which allows for custom responses inline. Fetch
+             * requests are mocked by fetchMock which requires initialization
+             * here.
+             */
+            fetchMock
+              .get(/include-me-header-test/, 200)
+              .get(/exclude-me-header-test/, 200);
+            sinon.useFakeXMLHttpRequest();
```

Comment:
this should not be here but in unit tests - before starting, together with restoring after the unit tests are finished
for example
https://sinonjs.org/releases/latest/fake-xhr-and-server/

Reply:
This is here for the same reason the `fetchMock` is. Because once the lightstep code is loaded it immediately grabs window to try and make sure that it's the first thing to get the `XMLHttpRequest` object. If later calls are made to `useFakeXMLHttpRequest`or `fetchMock` they don't do anything because lightstep has wrapped the real objects already.

Do you have any suggestions for workarounds? I considered writing a mock for wrapping these window functions and then either returning the mock, or not, but that requires some knowledge of exactly how these mocks are wrapping the window objects in the first place.

Basically I agree, just given the limitations and how the plugins are written as of today I couldn't come up with something that seemed both effective but also not overkill.

Reply:
@trescenzi if this is not possible then @mwear should decide. I think we should create a separate issue to move the fetchMock and sinon into appropriate unit tests. But this is really strange as it looks a bit for me like hacky solution as sinon or fetchMock should be used in a way that you can have full control over the tested class before and after the unit test so you can easily mock it or remove the mock. If this is not possible then something is wrong and should be fixed.

Reply:
If I'm not mistaken what's happening is that the [lightstep plugin](https://github.com/lightstep/lightstep-tracer-javascript/blob/master/src/plugins/instrument_fetch.js#L6) looks at the window fetch immediately and assigns it as the "proxied" fetch. If in the future something like `fetchMock` mocks fetch it'll grab `window.fetch` and replace it with the mock. However at that point lightstep has already stored its reference to the window's `fetch` which means that even though `fetch` is proxied lightstep will just call through to the old value.

I believe code like the following would work to manage this:
```JavaScript
window._fetch = window.fetch;
Object.define(window, 'fetch', {
  get: function() {
    return this._fetch;
  },
  set: function(newFetch) {
    this._fetch = newFetch;
  }
});
```
So even though lightstep would store a reference to `window.fetch` that reference would be "live" and *should* always call through to the correct, be it the real implementation or the mock one, at the time of usage. I'm going to try this out. It's hacky, but does feel better than just always mocking things.

Reply:
So with a quick test that makes `fetchMock` not work. I'm really not sure if there's a way to solve this without changing the lightstep code itself. Since it's storing a reference to whatever `window.fetch`/`window.XHMHttpRequest` is at the time it's loaded it makes it really rough to dynamically turn on/off anything that wraps those methods.

Reply:
So with a quick test that makes `fetchMock` not work. I'm really not sure if there's a way to solve this without changing the lightstep code itself. Since it's storing a reference to whatever `window.fetch`/`window.XHMHttpRequest` is at the time it's loaded it makes it really rough to dynamically turn on/off anything that wraps those methods.

Reply:
@trescenzi I understand and I'm fine with keeping it as you did :). But also we should have a new issue to make all what is necessary to be able to use test tools (sinon, fetchmock etc.) correctly inside unit tests, WDYT @mwear ?

Reply:
Ok cool. I'm largely just continuing to think about this because I agree with you. It's surprising that we can't use these mocks as expected since they are super common, standard solutions for testing network requests.

With regards to an issue I might actually be able to file an issue that's relevant to my usage of the plugins. When incorporating these plugins I actually broke my e2e tests because of a similar problem. I have to go back and look at what happened though to see if it's something that makes sense to be an issue.

Reply:
I'm fine with addressing the testing issues separately, as I would really like to get this work in so folks can use it.